File: org.boris.winrun4j/java/src/org/boris/winrun4j/EventLog.java
/*******************************************************************************
 * This program and the accompanying materials
 * are made available under the terms of the Common Public License v1.0
 * which accompanies this distribution, and is available at 
 * http://www.eclipse.org/legal/cpl-v10.html
 * 
 * Contributors:
 *     Peter Smith
 *******************************************************************************/
package org.boris.winrun4j;

/**
 * A mechanism for adding events.
 */
public class EventLog
{
    public static final int SUCCESS = 0x0000;
    public static final int ERROR = 0x0001;
    public static final int WARNING = 0x0002;
    public static final int INFORMATION = 0x0004;
    public static final int AUDIT_SUCCESS = 0x0008;
    public static final int AUDIT_FAILURE = 0x0010;

    private static long library = Native.loadLibrary("advapi32");
    
    /**
     * Report an event.
     * 
     * @param source.
     * @param type.
     * @param msg.
     * 
     * @return boolean.
     */
    public static boolean report(String source, int type, String msg) {
        long buf = NativeHelper.toNativeString(source, true);
        long h = NativeHelper.call(library, "RegisterEvent", 0, buf);
        long m = NativeHelper.toNativeString(msg, true);
        boolean res = NativeHelper.call(library, "ReportEvent", new long[] { h, type, 0, 0, 0, 0, msg.length() * 2, 0, m }) == 1;
        NativeHelper.free(buf, m);
        return res;
    }
}

File: WinRun4J/src/launcher/EventLog.cpp
/*******************************************************************************
* This program and the accompanying materials
* are made available under the terms of the Common Public License v1.0
* which accompanies this distribution, and is available at 
* http://www.eclipse.org/legal/cpl-v10.html
* 
* Contributors:
*     Peter Smith
*******************************************************************************/

#include "EventLog.h"
#include "../common/Log.h"
#include "../java/JNI.h"

bool EventLog::RegisterNatives(JNIEnv *env)
{
	Log::Info("Registering natives for EventLog class");
	jclass clazz = JNI::FindClass(env, "org/boris/winrun4j/EventLog");
	if(clazz == NULL) {
		Log::Warning("Could not find EventLog class");
		if(env->ExceptionCheck())
			env->ExceptionClear();
		return false;
	}

	JNINativeMethod methods[1];
	methods[0].fnPtr = (void*) Report;
	methods[0].name = "report";
	methods[0].signature = "(Ljava/lang/String;ILjava/lang/String;)Z";
	env->RegisterNatives(clazz, methods, 1);
	if(env->ExceptionCheck()) {
		JNI::PrintStackTrace(env);
		env->ExceptionClear();
	}

	return true;
}

bool EventLog::Report(JNIEnv* env, jobject self, jstring source, jint type, jstring msg)
{
	if(source == NULL || msg == NULL)
		return false;

	jboolean iscopy = false;
	const char* src = source ? env->GetStringUTFChars(source, &iscopy) : 0;
	const char* m = msg ? env->GetStringUTFChars(msg, &iscopy) : 0;

	HANDLE h = RegisterEventSource(0, src);
	if(h == NULL) {
		return false;
	}

	return ReportEvent(h, type, 0, 0, 0, 0, strlen(m), 0, (void *) m);
}
